src/west/manifest.py

```diff
@@ -436,6 +482,140 @@ def as_dict(self):
             ret['west-commands'] = self.west_commands
         return ret
 
+    def format(self, s, *args, **kwargs):
+        '''Calls s.format() with instance-related format keys.
+
+        The formatted value is returned.
+
+        :param s: string (or other object) whose format() method to call
+
+        The format method is called with *args and the following kwargs:
+
+        - this object's __slots__ / values (name, url, etc.)
+        - name_and_path: "self.name + (self.path)"
+        - remote_name: "None" if no remote, otherwise self.remote.name
+        - any additional kwargs passed as parameters
+
+        The kwargs passed as parameters override the other values.
+        '''
+        kwargs = self._format_kwargs(kwargs)
+        return s.format(*args, **kwargs)
+
+    def _format_kwargs(self, kwargs):
+        ret = {s: getattr(self, s) for s in self.__slots__}
+        ret['name_and_path'] = '{} ({})'.format(self.name, self.path)
+        ret['remote_name'] = ('None' if self.remote is None
+                              else self.remote.name)
+        ret.update(kwargs)
+        return ret
+
+    def git(self, cmd, extra_args=(), capture_stdout=False, check=True,
```

Comment:
Just picked a line here.

I thought we wanted to avoid having `git` knowledge in manifest.py.
https://github.com/zephyrproject-rtos/west/pull/189#issuecomment-461498125

If starting to refactor, then maybe then maybe we should take up #191 in order to place `git` handling correctly ;)

Reply:
> I thought we wanted to avoid having `git` knowledge in manifest.py.
> [#189 (comment)](https://github.com/zephyrproject-rtos/west/pull/189#issuecomment-461498125)

I wasn't against doing it, just against doing an unnecessary refactor during the stabilization period. It is necessary now in order to implement this bug fix, which has been marked high priority for the next release.

> If starting to refactor, then maybe then maybe we should take up #191 in order to place `git` handling correctly ;)

Please take a look at the commit message for the patch:

```
    More clean up for manifest.py / project.py responsibilities
    
    This commit is further progress on #191, but doesn't fix it entirely.
    
    West's design has always been to split code into reusable libraries
    and commands which wrap them. The extra effort is justified since West
    is pluggable by extension commands, which often benefit from being
    able to use the library pieces.
    
    Although the code in project.py started out self-contained, it's
    gradually being refactored to fit the design.
    
    Continue that by factoring two more features into manifest.py: helpers
    for running git commands in a project and formatting strings using
    project attributes.
    
    This lets us remove almost all of the wrappers around west.log in a
    way that makes the code easier to read. The remaining _inf() is
    renamed _banner() to make its purpose clearer.
    
    Beyond being a useful refactoring which makes extensions easier to
    write, this is needed to fix a bug where west doesn't check if the
    currently running version is not a descendant of the manifest
    revision, indicating it is out of date and prompting a warning. Fixing
    that is a high priority issue for Zephyr v1.14.
    
    Delete a fair bit of dead code left over from the older multirepo
    commands which were deleted from v0.4.0 while we are here.
```

I more cleanup is eventually warranted (I only factored out a few nondestructive git based operations for now) but I didn't want to do that without good reason outside of a merge window.

Reply:
> Please take a look at the commit message for the patch:

whoops. I see you do refer to #191, my bad :(

I was just under the impression you prefer to have `git` knowledge outside of `manifest.py` module, which is something I would agree with.

From: https://github.com/zephyrproject-rtos/west/pull/189#issuecomment-461498125
> If I understand what you're saying, then a problem with that approach is that that would require the manifest class to know how to run Git to freeze the revisions. That felt like an abstraction violation to me, which is why I chose to do it this way. 

So follow up question, do you think we should factor out the `git` operations into independent module ?
So that they can be imported by manifest.py and project.py ?

Reply:
Looking at the code, I think it reads pretty naturally when you can do `project.git_related_operation()`, so I'm actually thinking it's nice to have it in `manifest.py`.

Reply:
ok.
not so worreid here.

But maybe at later time we should consider a `git` module, so that we avoid duplicating same `git` commands in multiple modules in future.